File: framec_tests/golang/transition_params/transition_params.go
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files
package transition_params

import (
    "strconv"
    "golang/framelang")


func NewTransitParams() TransitParams {
    m := &transitParamsStruct{}
    
    // Validate interfaces
    var _ TransitParams = m
    var _ TransitParams_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewTransitParamsCompartment(TransitParamsState_Init)
    
    // Override domain variables.
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type TransitParamsState uint

const (
    TransitParamsState_Init TransitParamsState = iota
    TransitParamsState_A
    TransitParamsState_B
)

type TransitParams interface {
    Next() 
    Change() 
}

type TransitParams_actions interface {
    log(msg string) 
}


type transitParamsStruct struct {
    _compartment_ *TransitParamsCompartment
    _nextCompartment_ *TransitParamsCompartment
    tape []string
}

//===================== Interface Block ===================//

func (m *transitParamsStruct) Next()  {
    e := framelang.FrameEvent{Msg:"Next"}
    m._mux_(&e)
}

func (m *transitParamsStruct) Change()  {
    e := framelang.FrameEvent{Msg:"Change"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *transitParamsStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case TransitParamsState_Init:
        m._TransitParamsState_Init_(e)
    case TransitParamsState_A:
        m._TransitParamsState_A_(e)
    case TransitParamsState_B:
        m._TransitParamsState_B_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *transitParamsStruct) _TransitParamsState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "Next":
        compartment := NewTransitParamsCompartment(TransitParamsState_A)
        compartment.EnterArgs["msg"] = "hi A"
        m._transition_(compartment)
        
        return
    case "Change":
        compartment := NewTransitParamsCompartment(TransitParamsState_A)
        
        m._changeState_(compartment)
        
        return
    }
}

func (m *transitParamsStruct) _TransitParamsState_A_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.log(e.Params["msg"].(string))
        
        return
    case "<":
        m.log("bye A")
        
        return
    case "Next":
        compartment := NewTransitParamsCompartment(TransitParamsState_B)
        compartment.EnterArgs["msg"] = "hi B"
        compartment.EnterArgs["val"] = 42
        m._transition_(compartment)
        
        return
    case "Change":
        compartment := NewTransitParamsCompartment(TransitParamsState_B)
        
        m._changeState_(compartment)
        
        return
    }
}

func (m *transitParamsStruct) _TransitParamsState_B_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.log(e.Params["msg"].(string))
        m.log(strconv.Itoa(e.Params["val"].(int)))
        
        return
    case "<":
        m.log(strconv.FormatBool(e.Params["val"].(bool)))
        m.log(e.Params["msg"].(string))
        
        return
    case "Next":
        m._compartment_.ExitArgs["val"] = true
        m._compartment_.ExitArgs["msg"] = "bye B"
        compartment := NewTransitParamsCompartment(TransitParamsState_A)
        compartment.EnterArgs["msg"] = "hi again A"
        m._transition_(compartment)
        
        return
    case "Change":
        compartment := NewTransitParamsCompartment(TransitParamsState_A)
        
        m._changeState_(compartment)
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *transitParamsStruct) _transition_(compartment *TransitParamsCompartment) {
    m._nextCompartment_ = compartment
}

func (m *transitParamsStruct) _do_transition_(nextCompartment *TransitParamsCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

func (m *transitParamsStruct) _changeState_(compartment *TransitParamsCompartment) {
    m._compartment_ = compartment
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *transitParamsStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type TransitParamsCompartment struct {
    State TransitParamsState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewTransitParamsCompartment(state TransitParamsState) *TransitParamsCompartment {
    c := &TransitParamsCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/transition_params/transition_params.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters


class TransitParams:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = TransitParamsCompartment('__transitparams_state_Init', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = [None]
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def Next(self,):
        self.return_stack.append(None)
        __e = FrameEvent("Next",None)
        self.__kernel(__e)
        return self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $Init
    
    def __transitparams_state_Init(self, __e, compartment):
        if __e._message == "Next":
            next_compartment = None
            next_compartment = TransitParamsCompartment('__transitparams_state_A', next_compartment)
            next_compartment.enter_args["msg"] = "hi A"
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $A
    
    def __transitparams_state_A(self, __e, compartment):
        if __e._message == ">":
            self.log_do(__e._parameters["msg"])
            return
        elif __e._message == "<":
            self.log_do("bye A")
            return
        elif __e._message == "Next":
            next_compartment = None
            next_compartment = TransitParamsCompartment('__transitparams_state_B', next_compartment)
            next_compartment.enter_args["msg"] = "hi B"
            next_compartment.enter_args["val"] = 42
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $B
    
    def __transitparams_state_B(self, __e, compartment):
        if __e._message == ">":
            self.log_do(__e._parameters["msg"])
            self.log_do(str(__e._parameters["val"]))
            return
        elif __e._message == "<":
            self.log_do(str(__e._parameters["val"]))
            self.log_do(__e._parameters["msg"])
            return
        elif __e._message == "Next":
            self.__compartment.exit_args["val"] = True
            self.__compartment.exit_args["msg"] = "bye B"
            next_compartment = None
            next_compartment = TransitParamsCompartment('__transitparams_state_A', next_compartment)
            next_compartment.enter_args["msg"] = "hi again A"
            self.__transition(next_compartment)
            return
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__transitparams_state_Init':
            self.__transitparams_state_Init(__e, self.__compartment)
        elif self.__compartment.state == '__transitparams_state_A':
            self.__transitparams_state_A(__e, self.__compartment)
        elif self.__compartment.state == '__transitparams_state_B':
            self.__transitparams_state_B(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class TransitParamsCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    